src/main/java/cubicchunks/world/cube/Cube.java

```diff
@@ -422,6 +429,73 @@ public void randomTick(WorldServer worldServer, Random rand) {
     }
+    
+    public void updateThunder(WorldServer worldServer, Random rand, boolean doMobSpawning) {
+        this.updateLCG = this.updateLCG * 3 + 1013904223;
+        int j1 = updateLCG >> 2;
+        int localX = AddressTools.getLocalX(j1);
+        int localZ = AddressTools.getLocalZ(j1);
+        int localY = AddressTools.getLocalY(j1);
+        BlockPos randomPos = this.coords.localToBlock(localX, localY, localZ);
+        BlockPos strikeTarget = this.column.getPrecipitationHeight(randomPos);
+        if (Coords.blockToCube(strikeTarget.getY()) != this.getY())
+            return;
+        AxisAlignedBB aabb = (new AxisAlignedBB(strikeTarget, randomPos)).grow(3.0D);
+        Iterable<EntityLivingBase> setOfLiving = this.entities.getEntitySet().getByClass(EntityLivingBase.class);
+        for(EntityLivingBase entity:setOfLiving){
+            if(!entity.isEntityAlive())
+                continue;
+            BlockPos entityPos = entity.getPosition();
+            if (entityPos.getY() < this.column.getHeightValue(Coords.blockToLocal(entityPos.getX()), Coords.blockToLocal(entityPos.getZ())))
+                continue;
+            if(entity.getEntityBoundingBox().intersects(aabb)){
+                // This entity is lucky!
+                strikeTarget = entityPos;
+                break;
+            }
+        }
+        DifficultyInstance difficultyinstance = this.world.getDifficultyForLocation(strikeTarget);
+        if (doMobSpawning && rand.nextDouble() < (double) difficultyinstance.getAdditionalDifficulty() * 0.01D) {
+            EntitySkeletonHorse entityskeletonhorse = new EntitySkeletonHorse(worldServer);
+            entityskeletonhorse.setTrap(true);
+            entityskeletonhorse.setGrowingAge(0);
+            entityskeletonhorse.setPosition((double) strikeTarget.getX(), (double) strikeTarget.getY(), (double) strikeTarget.getZ());
+            this.world.spawnEntity(entityskeletonhorse);
+            worldServer.addWeatherEffect(new EntityLightningBolt(worldServer, (double) strikeTarget.getX(), (double) strikeTarget.getY(),
+                    (double) strikeTarget.getZ(), true));
+        } else {
+            worldServer.addWeatherEffect(new EntityLightningBolt(worldServer, (double) strikeTarget.getX(), (double) strikeTarget.getY(),
+                    (double) strikeTarget.getZ(), false));
+        }
+    }
+    
+    public void updateRainSnowIce(WorldServer worldServer, CubeProviderServer cubeProvider, Random rand, boolean isRaining) {
+        this.updateLCG = this.updateLCG * 3 + 1013904223;
+        int j1 = updateLCG >> 2;
+        int localX = AddressTools.getLocalX(j1);
+        int localZ = AddressTools.getLocalZ(j1);
+        int localY = AddressTools.getLocalY(j1);
+        BlockPos randomPos = this.coords.localToBlock(localX, localY, localZ);
+        BlockPos precipitationHeight = this.column.getPrecipitationHeight(randomPos);
+        if (Coords.blockToCube(precipitationHeight.getY()) != this.getY())
```

Comment:
Same applies here, all you need is to get precipitation height for each xz in Column, get cube for that height if loaded, and update snow there. No need to iterate over each cube. But in this case we are basically using vanilla code, with the isLoaded check.